### DISCUSSION TITLE
Passing an array to timeline_variables?

### DISCUSSION THREAD

**Discussion by morgan-patrick** [jspsych/jsPsych#2938] [Q&A] [answered]
Hello All!

I am currently trying to resolve an issue where jsPsych is not recognizing the stimulus objects passed into the jsPsychVideoButtonResponse plugin. The stimulus parameter (screenshot below) has the jsPsych.timelineVariable calling 'stimulus', which is the name of the stimulus objects listed in an array that is passed to timeline_variables. Has anyone else encountered an issue passing an array to timeline_variables? I cannot pass individual file names directly to this variable because they are shuffled and subsetted into that list. 

Grateful for any insight!
![2](https://user-images.githubusercontent.com/63110313/212418493-020dfb7c-7283-4c55-bffd-91b210bcd497.png)
![1](https://user-images.githubusercontent.com/63110313/212418502-1b07a8d0-86bf-44a4-a04e-2384863960c3.png)

**Comment by becky-gilbert** [comment]
Hi @patrick10326, I think the problem is that `timeline_variables` needs to be set up differently. Instead of passing the `timeline_variables` array into the same object that defines the trial (i.e. the one with `type: jsPsychVideoButtonResponse` and related parameters such as `stimulus`, `choices`, etc.), you need to _create two different objects_: one that defines the trial, and another that defines the timeline variables for that trial. The [timeline variables documentation page](https://www.jspsych.org/7.3/overview/timeline/#timeline-variables) has more explanation and examples.

So in your case it might look something like this:
```js
// define the video button response trial object, and insert timeline variables wherever needed
var video_trial = {
  type: jsPsychVideoButtonResponse,
  stimulus: jsPsych.timelineVariable('stimulus'),
  choices: ['1','2','3','4'],
  prompt: '<p>Which event sticks out the most?</p>',
  response_allowed_while_playing: false,
  response_ends_trial: true
};

// set up the timeline variables procedure, which includes the trial(s) to repeat and the timeline variables array
var block_1 = {
  timeline: [video_trial],
  timeline_variables: stimulus_list_1,
  randomize_order: true
};

// add the procedure to your experiment timeline
experiment_timeline.push(block_1);
```
BTW, another way to write this is to define the video trial directly inside the `timeline` parameter in your `block_1` procedure. This is equivalent to the version above, so it's just a matter of preference.
```js
// set up the timeline variables procedure, which includes the trial(s) to repeat and the timeline variables array
var block_1 = {
  timeline: [{
    type: jsPsychVideoButtonResponse,
    stimulus: jsPsych.timelineVariable('stimulus'),
    choices: ['1','2','3','4'],
    prompt: '<p>Which event sticks out the most?</p>',
    response_allowed_while_playing: false,
    response_ends_trial: true
  }],
  timeline_variables: stimulus_list_1,
  randomize_order: true
};

// add the procedure to your experiment timeline
experiment_timeline.push(block_1);
```
Can you let us know if that solves the problem for you?

**Reply by morgan-patrick** [reply]
Hi @becky-gilbert , Thank you very much for your prompt and thorough response! Grateful for your insight. Unfortunately, neither solution solves the issue. I had previously tried mimicking the examples you linked as well, but none of them pass an array to the timeline_variables parameter, which led me to believe that is where the issue is. Grateful if you have any other suggestions! Thank you again.

**Reply by becky-gilbert** [reply]
Hmm that's weird, the `timeline_variables` parameter should definitely be an array of objects. Something like this:
```js
var stimulus_list_1 = [
 {stimulus: ['video/fish.mp4']},
 {stimulus: ['video/horse.mp4']},
 {stimulus: ['video/turtle.mp4']}
];
```
Is that the format of your array?

**Reply by morgan-patrick** [reply]
Weird indeed! Yes, my array is formatted that way. (I've tried with and w/o brackets around the file names.) For instance:

`stims = [
  {stimulus: "1_per_ls_hl.mp4"},
  {stimulus: "1_per_ls_lh.mp4"}
  ]`

The other thing I should mention is that I am defining stimulus_list_1 first as an empty array, and then pushing one of the two stimuli above to it; and repeating that push ten times from 10 different arrays like the one above to arrive at the version of stimulus_list_1 that is passed to the timeline_variable parameter. Is there a way to inspect or return the stimulus_list_1 array after these pushes, to ensure it's in the appropriate form? (Apologies for my lack of JS fluency!)

**Reply by becky-gilbert** [reply]
Aha! In your `stimulus_list_1` array, can you try putting all of your `stimulus` values inside an _array_? This is a weird quirk with the video plugins - they need the `stimulus` parameter to be an array rather than a string to allow for passing multiple file formats (see [video button response plugin parameters in the docs](https://www.jspsych.org/7.3/plugins/video-button-response/#parameters)).

And as far as checking the format of your array that you're generating - yes that's a great idea. You can just print it to the browser's console after whatever you do to create it in the code, e.g.:
```js
var stimulus_list_1 = [];

// code to dynamically generate the array goes here...

console.log('stimulus list 1: ', stimulus_list_1);
```

**Reply by morgan-patrick** [reply]
Hi Becky,

Thanks for your follow-up. If I understand you correctly, the stimulus_list_1 array needs to be of the form 
`stimulus_list_1 = ["file1.mp4", "file2.mp4"]`
and not 
`stimulus_list_1 = [{stimulus: "file1.mp4"}, {stimulus: "file2.mp4}].`
If that is correct, and the timeline_variables parameter below gets stimulus_list_1, then what key do we pass to the stimulus parameter?:
`var block_1 = {
  timeline: [{
    type: jsPsychVideoButtonResponse,
    stimulus: jsPsych.timelineVariable(**???**),
    choices: ['1', '2', '3', '4'],
    prompt: "<p>Which event sticks out the most?</p>",
    response_ends_trial: true,
    response_allowed_while_playing: false
    }],
  timeline_variables: stimulus_list_1,
  randomize_order: true

};`

Specifying `stimulus_list_1 = ["file1.mp4", "file2.mp4", "file3.mp4"]` and naming stimulus_list_1 as both the stimulus and timeline_variables of the plugin results in a successful presentation of the very first stimulus file, but this doesn't loop through all the files. Very grateful for any tips you might be able to provide!

**Comment by morgan-patrick** [comment, marked as answer]
@becky-gilbert I found the solution! I needed to place each stimulus file inside brackets in addition to having the entire array inside brackets as well. Thank you so much for all of this support! I'm grateful for your kindness!

**Reply by becky-gilbert** [reply]
No problem, glad you figured it out and thanks for letting us know!
